[12.0][ADD] purchase_stock_picking_invoice_link: linking purchases, p…

diggy128 opened the pull request:
…ickings and invoices

This is a port of #606 to v12.0
Fixes [issue](https://github.com/OCA/stock-logistics-workflow/pull/606#discussion_r398864179) pointed out by @omar7r 
[This issue](https://github.com/OCA/stock-logistics-workflow/pull/606#discussion_r398864179) seems invalid as invoice_line_ids do come formatted as [(6, 0, ids)] and the code runs.

@pedrobaeza regarding #500 for v11.0, it seems in v12.0 chained onchange on purchase_id is called as it should when using the purchase selector (vendor_bill_purchase_id) and links are created.

diggy128 committed:
[12.0][ADD] purchase_stock_picking_invoice_link: linking purchases, pickings and invoices

rousseldenis commented on purchase_stock_picking_invoice_link/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+
+{
+    'name': 'Purchase Stock Picking Invoice Link',
+    'version': '11.0.1.0.0',
```
Please change to 12...

rousseldenis commented on purchase_stock_picking_invoice_link/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+{
+    'name': 'Purchase Stock Picking Invoice Link',
+    'version': '11.0.1.0.0',
+    'category': 'Warehouse Management',
```
Please change to 'Warehouse'

rousseldenis commented on purchase_stock_picking_invoice_link/models/stock_move.py:
```diff
@@ -0,0 +1,18 @@
+    _inherit = "stock.move"
+
+    @api.multi
+    def _get_moves(self):
```
This method name is too general as it can exist in another module.

Maybe _get_purchase_invoice_link_moves()

rousseldenis commented on purchase_stock_picking_invoice_link/models/account_invoice.py:
```diff
@@ -0,0 +1,47 @@
+        for item in vals_list.get('invoice_line_ids'):
+            if not item[2].get('purchase_line_id'):
+                continue
+            moves = self.env['stock.move'].search([
```
Please build purchase_line_ids.ids list to make one search for performances

rousseldenis commented on purchase_stock_picking_invoice_link/models/account_invoice.py:
```diff
@@ -0,0 +1,47 @@
+            if moves:
+                move_ids = moves._get_moves()
+                item[2]['move_line_ids'] = [(6, 0, move_ids.ids)]
+                pickings |= move_ids.mapped('picking_id')
```
Experience shows that mapped is not really fast.
Use instead a set of ids and union() then browse all ids once.

rousseldenis commented on purchase_stock_picking_invoice_link/models/account_invoice.py:
```diff
@@ -0,0 +1,47 @@
+        res = super().create(vals_list)
+        if res:
+            for picking in pickings:
+                picking.invoice_ids = [(4, res.id)]
```
This assertion is not correct if vals_list is > 1

rousseldenis reviewed (changes requested):
Hi @diggy128 thanks for your contribution. Some remarks.

dalonsod commented:
@diggy128 any news on this? Thanks!

diggy128 commented:
Currently I won't be able to work into this as there are more important things in line and this works for me. However I keep the branch and will fix and commit once I've addressed these issues.

pedrobaeza commented:
Closing as not finished. 13.0 version in #883
